### PR TITLE
feat(auth): entity level authentication

### PR DESCRIPTION
**Pull request by meskill**
**Summary:**  

**Build & Testing:**

- [ ] I ran `cargo test` successfully.
- [ ] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [ ] I have performed a self-review of my code.
- [ ] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

## Summary by CodeRabbit

- **New Features**
	- Introduced and integrated comprehensive support for basic and JWT authentication methods.
	- Added GraphQL schemas for authentication settings and data retrieval with authentication.
	- Enhanced configuration handling to include authentication settings and protection levels for data.
- **Improvements**
	- Streamlined HTTP handling to accommodate authentication contexts and JSON response modifications.
	- Updated lambda expressions and path handling to support authentication validation and protected expressions.
- **Bug Fixes**
	- Fixed issues related to string borrowing across the application by utilizing `Cow<'_, str>` for more efficient string handling.
- **Tests**
	- Added a suite of test scenarios for basic and JWT authentication methods, covering various configurations and access control for protected data.
- **Documentation**
	- Included new documentation for configuring and utilizing authentication within the application.

### CODE REVIEW

File: examples/auth.graphql [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      }
+    ]
+    auth: [
+      {id: "basic-1", basic: {htpasswd: "{{vars.BASIC_AUTH}}"}}
+      {
+        id: "jwt-1"
+        jwt: {
+          jwks: {
+            remote: {
+              url: "https://gist.githubusercontent.com/meskill/b0e2343c4aa3e0af95ca0c9209b5818a/raw/tailcall-auth-jwt-jwks-example.json"
```
Comment (tusharmath) [inline examples/auth.graphql:22]: JWKS should be another type of link.

File: tests/execution/auth-jwt-const-env.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
+# Auth with JWT loaded from const env
+
+```graphql @server
+schema @server(port: 8000, graphiql: true, auth: [{jwt: {jwks: {data: "{{env.JWKS}}"}}, id: "jwt"}]) {
```
Comment (tusharmath) [inline tests/execution/auth-jwt-const-env.md:4]: Simply link the JWKS and HTPASSWD files and Tailcall should figure out what and how to authenticate.

File: tests/execution/auth-basic-file.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
+# Auth with BasicAuth loaded from const env
+
+```graphql @server
+schema
+  @server(port: 8000, graphiql: true, auth: [{id: "basic", basic: {htpasswd: "{{link.htpasswd}}"}}])
```
Comment (tusharmath) [inline tests/execution/auth-basic-file.md:5]:
```suggestion
  @server(port: 8000, graphiql: true)
```

Drop auth setting from server. We can implicitly understand that linking to an htpasswd file means the config requires authentication.

File: src/http/request_handler.rs [hunk 4]
```diff
@@ -74,10 +74,12 @@ fn not_found() -> Result<Response<Body>> {
...
     let upstream = app_ctx.blueprint.upstream.clone();
     let allowed = upstream.allowed_headers;
-    let req_headers = create_allowed_headers(req.headers(), &allowed);
+    let allowed_headers = create_allowed_headers(req.headers(), &allowed);
 
     let _allowed = app_ctx.blueprint.server.get_experimental_headers();
-    RequestContext::from(app_ctx).request_headers(req_headers)
+    RequestContext::from(app_ctx)
+        .request_headers(req.headers().clone())
+        .allowed_headers(allowed_headers)
```
Comment (tusharmath) [inline src/http/request_handler.rs:82]: This is more work than what we are doing in main (cloning all the request headers). We can keep a whitelist of headers or have some default "allowed-headers". In either case we should only keep a minimal set of headers in request context.

File: src/config/link.rs [hunk 5]
```diff
@@ -13,6 +13,7 @@ pub enum LinkType {
     Cert,
     Key,
     Operation,
+    File,
```
Comment (tusharmath) [inline src/config/link.rs:16]: Need HtPasswd and Jwks

File: src/config/config_module.rs [hunk 6]
```diff
@@ -35,6 +36,8 @@ impl<A> Deref for Content<A> {
 /// an IO operation, i.e., reading a file, making an HTTP call, etc.
 #[derive(Clone, Debug, Default)]
 pub struct Extensions {
+    pub files: HashMap<String, String>,
```
Comment (tusharmath) [inline src/config/config_module.rs:39]: don't need this.

**Comment by meskill** [comment]
Created new pr - https://github.com/tailcallhq/tailcall/pull/1579

**Pull request closed without merging**